pkg/syncer/upsync/upsync_controller.go

```diff
@@ -0,0 +1,174 @@
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/dynamic/dynamicinformer"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+)
+
+const (
+	controllerName = "kcp-resource-upsycner"
+	labelKey       = "kcp.resource.upsync"
+	syncValue      = "sync"
+)
+
+func isResourceToBeSynced(resource *unstructured.Unstructured) bool {
+	resourceLabels := resource.UnstructuredContent()["labels"].(labels.Set)
+	for key, value := range resourceLabels {
+		if key == labelKey && value == syncValue {
+			return true
+		}
+	}
+	return false
+}
+
+type Controller struct {
+	queue                     workqueue.RateLimitingInterface
+	upstreamClient            dynamic.ClusterInterface
+	downstreamClient          dynamic.Interface
+	downstreamNamespaceLister cache.GenericLister
+
+	syncerInformers     resourcesync.SyncerInformerFactory
+	syncTargetName      string
+	syncTargetWorkspace logicalcluster.Name
+	syncTargetUID       types.UID
+	syncTargetKey       string
+}
+
+type queueKey struct {
+	gvr schema.GroupVersionResource
+	key string
+}
+
+func (c *Controller) AddToQueue(gvr schema.GroupVersionResource, obj interface{}, logger logr.Logger) {
+	key, err := keyfunctions.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	logging.WithQueueKey(logger, key).V(2).Info("queueing GVR", "gvr", gvr.String())
+	c.queue.Add(
+		queueKey{
+			gvr: gvr,
+			key: key,
+		},
+	)
+}
+
+func NewUpSyncer(syncerLogger logr.Logger, syncTargetWorkspace logicalcluster.Name,
+	syncTargetName, syncTargetKey string, upstreamClient dynamic.ClusterInterface,
+	downstreamClient dynamic.Interface, upstreamInformers,
+	downstreamInformers dynamicinformer.DynamicSharedInformerFactory,
+	syncerInformers resourcesync.SyncerInformerFactory, syncTargetUID types.UID) (*Controller, error) {
+	c := &Controller{
+		queue:                     workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), controllerName),
+		upstreamClient:            upstreamClient,
+		downstreamClient:          downstreamClient,
+		downstreamNamespaceLister: downstreamInformers.ForResource(schema.GroupVersionResource{Version: "v1", Resource: "namespaces"}).Lister(),
+
+		syncerInformers:     syncerInformers,
+		syncTargetName:      syncTargetKey,
+		syncTargetWorkspace: syncTargetWorkspace,
+		syncTargetUID:       syncTargetUID,
+		syncTargetKey:       syncTargetKey,
+	}
+	logger := logging.WithReconciler(syncerLogger, controllerName)
+
+	syncerInformers.AddDownstreamEventHandler(
+		func(gvr schema.GroupVersionResource) cache.ResourceEventHandler {
+			logger.V(2).Info("Set up downstream informer", "gvr", gvr.String())
+			return cache.ResourceEventHandlerFuncs{
+				AddFunc: func(obj interface{}) {
+					if isResourceToBeSynced(obj.(*unstructured.Unstructured)) {
```

Comment:
Filtering in an event handler makes your controller edge-based, not level-driven. The state of the object at enqueue time does not necessarily match the state at reconcile time. Either filter via client options in the informer factory, or do it in reconciliation.